lightning/src/routing/router.rs

```diff
@@ -150,39 +159,180 @@ impl cmp::PartialOrd for RouteGraphNode {
+	/// This is useful for update_value_and_recompute_fees to make sure
+	/// we don't fall below the minimum. Should not be updated manually and
+	/// generally should not be accessed.
+	htlc_minimum_msat: u64,
+}
+
+impl PathBuildingHop {
+	// What we are paying for the use of this channels and the following channels.
+	fn get_fee_paid_msat(&self) -> u64 {
+		if let Some(fee_paid_msat) =
+		   self.next_hops_fee_msat.checked_add(self.hop_use_fee_msat) {
+			return fee_paid_msat;
+		} else {
+			return u64::max_value();
+		}
+	}
+}
+
+// Instantiated with a list of hops with correct data in them collected during path finding,
+// an instance of this struct should be further modified only via given methods.
+#[derive(Clone)]
+struct PaymentPath {
+	hops: Vec<PathBuildingHop>,
+}
+
+impl PaymentPath {
+
+	// TODO: Add a value_msat field to PaymentPath and use it instead of this function.
+	fn get_value_msat(&self) -> u64 {
+		self.hops.last().unwrap().route_hop.fee_msat
+	}
+
+	fn get_total_fee_paid_msat(&self) -> u64 {
+		if self.hops.len() < 1 {
+			return 0;
+		} else {
+			return self.hops.first().unwrap().next_hops_fee_msat;
+		}
+	}
+
+	// If an amount transferred by the path is updated, the fees should be adjusted.
+	// Any other way to change fees may result in an inconsistency.
+	// Also, it's only safe to reduce the value, to not violate limits like
+	// htlc_minimum_msat.
+	fn update_value_and_recompute_fees(&mut self, value_msat: u64) {
+		if value_msat == self.hops.last().unwrap().route_hop.fee_msat {
+			// Nothing to change.
+			return;
+		}
+		assert!(value_msat < self.hops.last().unwrap().route_hop.fee_msat);
+
+		let mut total_fee_paid_msat = 0 as u64;
+		for i in (1..self.hops.len()).rev() {
+			let cur_hop_amount_msat = total_fee_paid_msat + value_msat;
+			let mut cur_hop = self.hops.get_mut(i).unwrap();
+			cur_hop.next_hops_fee_msat = total_fee_paid_msat;
+			if let Some(new_fee) = compute_fees(cur_hop_amount_msat, cur_hop.channel_fees) {
+				cur_hop.hop_use_fee_msat = new_fee;
+				total_fee_paid_msat += new_fee;
+			} else {
+				// It should not be possible because this function is called only to reduce the
+				// value. In that case, compute_fee was already called with the same fees for
+				// larger amount and there was no overflow.
+				unreachable!();
+			}
+		}
+
+		// Propagate updated fees for the use of the channels to one hop back,
+		// where they will be actually paid (fee_msat).
+		// For the last hop it will represent the value being transferred over this path.
+		for i in 0..self.hops.len() - 1 {
+			let mut next_hop_use_fee_msat = self.hops.get(i + 1).unwrap().hop_use_fee_msat;
+			let mut cur_hop = self.hops.get_mut(i).unwrap();
+			let transferred_amount_msat = cur_hop.next_hops_fee_msat + value_msat + next_hop_use_fee_msat;
+			// Overpay in fees if we can't save these funds due to htlc_minimum_msat.
+			if let Some(extra_fees_msat) = cur_hop.htlc_minimum_msat.checked_sub(transferred_amount_msat) {
```

Comment:
I think overpay-fees-due-to-htlc-minimum-msat should be accounted somewhere in route scoring algorithm, or maybe have `htlc_minimum_msat` replace `base_fee` if higher...Otherwise a routing node could silently inflate its fees. Can you leave a TODO somewhere to mark this subtlety ?

Reply:
Absolutely.

Reply:
I actually added some simple prevention for this, and explained the whole issue. No test though, but the existing tests pass :)